File: backend/internal/model/pool_match_format.go
package model

const (
	MatchFormatLegacy = "legacy"
	MatchFormatFree   = "free"
	MatchFormatRaceTo = "race_to"

	PoolMatchMaxTargetWins      = 65
	PoolMatchMaxCompletedRounds = 129
)

func IsPoolMatchFormatGameType(gameType int) bool {
	return gameType == 3 || gameType == 4
}

func NormalizePoolMatchFormat(gameType int, format string, targetWins int) (string, int, bool) {
	if !IsPoolMatchFormatGameType(gameType) {
		if format == "" || format == MatchFormatLegacy {
			return MatchFormatLegacy, 0, targetWins == 0
		}
		return "", 0, false
	}

	switch format {
	case MatchFormatFree:
		return MatchFormatFree, 0, targetWins == 0
	case MatchFormatRaceTo:
		return MatchFormatRaceTo, targetWins, targetWins >= 1 && targetWins <= PoolMatchMaxTargetWins
	case "", MatchFormatLegacy:
		return MatchFormatLegacy, 0, targetWins == 0
	default:
		return "", 0, false
	}
}

func IsFlexiblePoolMatch(match *Match) bool {
	if match == nil || !IsPoolMatchFormatGameType(match.GameType) {
		return false
	}
	format, _, ok := NormalizePoolMatchFormat(match.GameType, match.MatchFormat, match.TargetWins)
	return ok && (format == MatchFormatFree || format == MatchFormatRaceTo)
}

func PoolMatchTargetReached(match *Match) bool {
	if !IsFlexiblePoolMatch(match) {
		return false
	}
	format, targetWins, ok := NormalizePoolMatchFormat(match.GameType, match.MatchFormat, match.TargetWins)
	return ok && format == MatchFormatRaceTo && targetWins > 0 &&
		(match.MyScore >= targetWins || match.OpponentScore >= targetWins)
}

File: app/utils/pool-match-format.js
export const POOL_MATCH_FORMAT_LEGACY = 'legacy'
export const POOL_MATCH_FORMAT_FREE = 'free'
export const POOL_MATCH_FORMAT_RACE_TO = 'race_to'
export const POOL_MATCH_MAX_ROUNDS = 129
export const POOL_MATCH_MAX_TARGET_WINS = 65
export const POOL_TARGET_OPTIONS = Array.from({ length: POOL_MATCH_MAX_TARGET_WINS }, (_, index) => index + 1)

export const isPoolMatchFormatGameType = (gameType) => [3, 4].includes(Number(gameType || 0))

export const normalizePoolMatchFormat = (payload = {}) => {
  const rawFormat = payload.match_format ?? payload.matchFormat ?? ''
  if (rawFormat === POOL_MATCH_FORMAT_LEGACY) {
    return { format: POOL_MATCH_FORMAT_LEGACY, targetWins: 0 }
  }
  const format = rawFormat === POOL_MATCH_FORMAT_RACE_TO
    ? POOL_MATCH_FORMAT_RACE_TO
    : POOL_MATCH_FORMAT_FREE
  const target = Number(payload.target_wins ?? payload.targetWins ?? 0) || 0
  return {
    format,
    targetWins: format === POOL_MATCH_FORMAT_RACE_TO
      ? Math.min(Math.max(target, 1), POOL_MATCH_MAX_TARGET_WINS)
      : 0
  }
}

export const getPoolMatchFormatLabel = (format, targetWins = 0) => {
  if (format === POOL_MATCH_FORMAT_LEGACY) return '传统赛制'
  return format === POOL_MATCH_FORMAT_RACE_TO
    ? `抢${Math.min(Math.max(Number(targetWins) || 1, 1), POOL_MATCH_MAX_TARGET_WINS)}局制`
    : '自由局数'
}

export const getPoolMatchFormatHint = (format, targetWins = 0) => {
  if (format === POOL_MATCH_FORMAT_LEGACY) return '按旧版手动结束规则进行'
  if (format !== POOL_MATCH_FORMAT_RACE_TO) return `不预设胜局，最多进行${POOL_MATCH_MAX_ROUNDS}局，允许平局`
  const target = Math.min(Math.max(Number(targetWins) || 1, 1), POOL_MATCH_MAX_TARGET_WINS)
  return `先达到${target}胜，最多可能进行${target * 2 - 1}局`
}

export const canFinishFreePoolMatch = ({ format, myScore, opponentScore } = {}) => {
  return format === POOL_MATCH_FORMAT_FREE && Number(myScore || 0) + Number(opponentScore || 0) >= 1
}

export const canContinuePoolMatch = ({ format, targetWins, myScore, opponentScore } = {}) => {
  const myWins = Number(myScore) || 0
  const opponentWins = Number(opponentScore) || 0
  if (format === POOL_MATCH_FORMAT_FREE) return myWins + opponentWins < POOL_MATCH_MAX_ROUNDS
  if (format === POOL_MATCH_FORMAT_RACE_TO) {
    const target = Math.min(Math.max(Number(targetWins) || 1, 1), POOL_MATCH_MAX_TARGET_WINS)
    return myWins < target && opponentWins < target
  }
  return true
}
